docs/architecture/adr-001-umee-module.md

```diff
@@ -0,0 +1,90 @@
+
+## Changelog
+
+- April 28, 2021: Initial Draft (@alexanderbez)
+
+## Status
+
+Proposed
+
+## Context
+
+Umee is a Universal Capital Facility that can collateralize assets on one
+blockchain towards borrowing assets on another blockchain. The platform specializes
+in allowing staked assets from POS blockchains to be used as collateral for
+borrowing across blockchains. The platform uses a combination of algorithmically
+determined interest rates based on market driven conditions.
+
+For the initial MVP implementation of the Umee network, we require the ability
+for users to be able to send ATOM tokens to a dedicated pool in the Umee network
+via IBC. ATOM tokens deposited into the Umee pool will mint a derivative meToken
+in a one-to-one ratio, i.e. one ATOM mints one meToken derivative.
+
+Validators on the Umee network will take these deposited ATOM tokens and delegate
+them to a set of governance-controlled validators on the ATOM source chain,
+e.g. the Cosmos Hub.
+
+A user that sent ATOM tokens to the Umee network can then take their derivative
+meTokens and send them to Ethereum via a bridge where a synthetic ERC-20 version
+of the meTokens are minted. Once sent, the meTokens are locked in Umee and the
+user can then freely trade and operate within Ethereum's DeFi ecosystem with the
+synthetic ERC-20 meTokens.
+
+## Decision
+
+To support users sending ATOM tokens to the Umee network via IBC, we will create
+a `x/umee` module that contains a dedicated `ModuleAccount` that represents the
+ATOM pool. In addition, we will implement a [custom IBC module](https://github.com/cosmos/ibc-go/blob/v1.0.0-alpha1/docs/custom.md).
+The custom IBC module will handle receiving custom packets on a custom Umee port.
+The custom module will be nearly identical to the ICS-20 module, except that it
+will perform the following:
+
+- Automatically send source ATOM tokens to the dedicated `ModuleAccount` pool
+  address instead of a user supplied address.
+- Mark the transfer's amount in a pending queue, `PendingDelegationQueue`, to be
+  later used for delegation on the source chain.
+- Track the sender and amount in the module's state.
+- Mint derivative meTokens and send them to the source sender's Umee account.
+
+At the beginning of a staking epoch, E<sub>1</sub> in `BeginBlock`, the `x/umee`
+module will perform the following:
+
+- Move the funds from the `PendingDelegationQueue` into another queue,
+  `PrevPendingDelegationQueue`. This queue is locked as funds cannot be sent to
+  it. In addition, these are the funds that will be delegated to the source chain
+  during the current staking epoch.
+- Clear out `PendingDelegationQueue`.
+
+The validators in the current staking epoch, E<sub>1</sub>, will then construct,
+sign, and broadcast multisig transactions to the Umee network. Once enough
+transactions are received by the `x/umee` module, specifically 2/3 of total
+staking weight, the `x/umee` module will store the constructed multisig transaction
```

Comment:
staking weight of the validators signers of the multisig? or the total power? Can a validator not be part of the multisig?

Reply:
Staking weight of the validators in the validator set, i.e. of total power. All validators have to be in the multisig per epoch. You cannot opt out.

Reply:
ack, it would be good to write that in the ADR

Reply:
So when the validator joins, the multisig weights will have to be updated with a weighted multisig design. This seems really hard to do given that there are multisigs on every chain Umee supports.

It would be interesting to have a design that allows you to put a weighted multisig on Umee, but then proxies everything to other accounts on other chains. The concept would be:
- On Umee, I have a weighted multisig
- On Cosmos, I have an account that when it receives a threshold signature from the weighted multisig originated from Umee chain, it performs the action I want (i.e. Delegation)

Reply:
There are no weights. If the account already exists on the native ATOM chain, great, we do nothing. If it does not exist, it'll have to be created first, where the Umee network will send a tx to the native ATOM chain to create the account (by sending some funds to it). The multi-sig is not weighted, it's a simple n/n setup where all validators are required to sign.

Reply:
It seems like having a weighted multisig would be worth it and the sdk would have to be extended to support it - otherwise it would limit who can become a validator on the network

Reply:
There is no limit to who can become a validator on Umee, but having power-based multi-sig would be nice.